### PR TITLE
#1569 DTO pattern implementation using Enums

### PR DESCRIPTION
**Pull request by siavashsoleymani**
Pull request title

- This PR implements a DTO design pattern using java Enums.
[Related issue](https://github.com/iluwatar/java-design-patterns/issues/1569 )

Pull request description

- The new module added in pom.xml for this different implementation of the DTO design pattern
- DTO design pattern using Enums implemented
- Google code style check passed
- Test case implemented just for running the application to see it's functionality.

> For detailed contributing instructions see https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

### CODE REVIEW

File: data-transfer-object-enum-impl/pom.xml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                                    <mainClass>com.iluwatar.datatransferenum.App</mainClass>
+                                </manifest>
+                            </archive>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (ohbus) [inline data-transfer-object-enum-impl/pom.xml:37]: add an extra line

File: data-transfer-object-enum-impl/etc/data-transfer-object-enum-impl.urm.puml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+Create ..|> Cost 
+Create ..|> Supplier 
+Private ..|> Id 
+Private ..|> Name 
+Private ..|> Price 
+Private ..|> Cost 
+Public ..|> Id 
+Public ..|> Name 
+Public ..|> Price 
+@enduml
```
Comment (ohbus) [inline data-transfer-object-enum-impl/etc/data-transfer-object-enum-impl.urm.puml:129]: add an extra line

**Review by ohbus — CHANGES_REQUESTED** [review]

File: data-transfer-object-enum-impl/etc/data-transfer-object-enum-impl.urm.puml (hunk 2 above)
Comment (siavashsoleymani) [reply data-transfer-object-enum-impl/etc/data-transfer-object-enum-impl.urm.puml:129]: Done, thanks for your suggestions

**Comment by siavashsoleymani** [comment]
@ohbus would you please review this?

**Comment by siavashsoleymani** [comment]
@iluwatar would you please review this? I understand that you might busy but I think it is worth to be merge/hacktoberfest-accept in this good repo.

**Comment by ravening** [comment]
To be frank, I dont see any advantage in using this pattern. It can be achieved with regular classes and feels like you are just complicating it with enums and other things.

Instead of KISS(keep it simple silly), you are complicating yourself

**Comment by siavashsoleymani** [comment]
You can find the advantages of using this pattern in the link below:

https://blog.scottlogic.com/2020/01/03/rethinking-the-java-dto.html

Also I'm not agree with you that is not simple. Probably because it's something new and is not familiar with our mindset can cause this thought.

**Comment by ravening** [comment]
> You can find the advantages of using this pattern in the link below:
> 
> https://blog.scottlogic.com/2020/01/03/rethinking-the-java-dto.html
> 
> Also I'm not agree with you that is not simple. Probably because it's something new and is not familiar with our mindset can cause this thought.

Actually, this is not new. DTO has been in place for many years and can be easily achieved with classes. If you want to complicate simple stuff then go ahead :D :D

**Comment by siavashsoleymani** [comment]
:))) by new I mean it's not familiar to your eyes not the DTO pattern itself:))

Anyway, everyone has an opinion, and of course, their opinion is respected

File: data-transfer-object-enum-impl/src/main/java/com/iluwatar/datatransferenum/Product.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    this.supplier = supplier;
+  }
+
+  /**
+   * Constructor.
+   */
+  public Product() {
+  }
+
+  public Long getId() {
```
Comment (ravening) [inline data-transfer-object-enum-impl/src/main/java/com/iluwatar/datatransferenum/Product.java:36]: A suggestion. You can use Lombok packages from this repo which avoids all the boilerplate code

**Review by ohbus — APPROVED** [review]

**Comment by ohbus** [comment]
@siavashsoleymani Thank you for your contribution to this repository. You have done a good job. 🎉
Please go the **issue** #1569 and provide a solid reference in the issue description. _Treat this as important._

@all-contributors please add @siavashsoleymani for code

**Comment by iluwatar** [comment]
I see a couple of problems with the implementation.
- It has been created in its own folder like a new pattern although we already have `data-transfer-object`. Due to that it's duplicating information in many places including `README.md` and can't be differentiated on the web site
- I'm not sure the test coverage is sufficient. This one only has `AppTest.java`
- It's not using local variable type inference (var)
I created a new ticket https://github.com/iluwatar/java-design-patterns/issues/1617 to refactor this.

File: data-transfer-object-enum-impl/src/main/java/com/iluwatar/datatransferenum/Product.java (hunk 3 above)
Comment (dvsingh9) [reply data-transfer-object-enum-impl/src/main/java/com/iluwatar/datatransferenum/Product.java:36]: agree with you @ravening

**Pull request merged**